src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -45,29 +46,45 @@ private abstract class Accessor<TKey, TWriteQueueKey, TDatabaseId>
             private readonly string _delete_from_writecache_table;
             private readonly string _insert_or_replace_into_main_table_select_star_from_writecache_table;
 
-            public Accessor(SQLitePersistentStorage storage)
+            public Accessor(
+                SQLitePersistentStorage storage,
+                ImmutableArray<(string name, string type)> primaryKeys)
             {
+                Storage = storage;
+                _primaryKeys = primaryKeys;
+
                 var main = Database.Main.GetName();
                 var writeCache = Database.WriteCache.GetName();
 
-                var dataTableName = this.Table switch
-                {
-                    Table.Solution => SolutionDataTableName,
-                    Table.Project => ProjectDataTableName,
-                    Table.Document => DocumentDataTableName,
-                    _ => throw ExceptionUtilities.UnexpectedValue(this.Table),
-                };
+                // "X" = ? and "Y" = ? and "Z" = ?
+                var whereClause = string.Join(" and ", primaryKeys.Select(k => $@"{k.name} = ?"));
+
+                var dataTableName = this.TableName;
+                _select_rowid_from_main_table_where_0 = $"select rowid from {main}.{dataTableName} where {whereClause}";
+                _select_rowid_from_writecache_table_where_0 = $"select rowid from {writeCache}.{dataTableName} where {whereClause}";
+
+                var allColumnNames = primaryKeys.Select(t => t.name).Concat(ChecksumColumnName).Concat(DataColumnName);
+
+                _insert_or_replace_into_writecache_table_values_0_1_2 = $@"insert or replace into {writeCache}.{dataTableName}
+                    ({string.Join(",", allColumnNames)})
+                    values
+                    ({string.Join(",", allColumnNames.Select(n => "?"))})";
```

Comment:
ugly manual generation (which we were doing before, but which gets even uglier now).  but this is only executed once in the lifetime of the host session.  we control all data here (like the column names) so it's not terrible.